### PR TITLE
Ready: Fix BinaryValue json encoding for MapReduce inputs

### LINKED ISSUE #520: Wrong JSON format for IndexMapReduce [JIRA: CLIENTS-413]
I'm trying to perform a simple mapreduce query with 2.0.0 against Riak 2.0.0 (from Scala):

``` scala
    val query: IndexMapReduce = (new IndexMapReduce.Builder)
      .withNamespace(new Namespace("type", "Bucket"))
      .withIndex("rawId_bin")
      .withMatchValue(BinaryValue.createFromUtf8("06403f5efcc17e4a"))
      .withReducePhase(RiakFunction.newErlangFunction("riak_kv_mapreduce", "reduce_count_inputs"))
      .build()
```

This fails with

```
   {error,function_clause,
   [{riak_index,parse_binary,
                [{struct,[{<<"value">>,<<"06403f5efcc17e4a">>}]}],
                [{file,"src/riak_index.erl"},{line,543}]},
    {riak_index,parse_field,3,[{file,"src/riak_index.erl"},{line,192}]},
    {riak_index,'-parse_fields/1-fun-0-',3,
                [{file,"src/riak_index.erl"},{line,158}]},
    {lists,foldl,3,[{file,"lists.erl"},{line,1248}]},
    {riak_index,parse_fields,1,[{file,"src/riak_index.erl"},{line,167}]},
    {riak_kv_mapred_json,parse_index_input,1,
                         [{file,"src/riak_kv_mapred_json.erl"},
                          {line,231}]},
    {riak_kv_mapred_json,parse_request,1,
                         [{file,"src/riak_kv_mapred_json.erl"},{line,55}]},
    {riak_kv_wm_mapred,verify_body,2,
                       [{file,"src/riak_kv_wm_mapred.erl"},{line,170}]}]}
```

Apparently, the MatchCriteria part of the request is malformed: Riak expects it to be `"key": "06403f5efcc17e4a"` but the actual request contains `"key": { "value": "06403f5efcc17e4a" }`

### PR DESCRIPTION
**Pull request by alexmoore**
Fixes #520 (CLIENTS-413).

### CODE REVIEW

**Comment by srgg** [comment]
@alexmoore why don't us get rid of BinaryValue as we did for TS?

File: src/main/java/com/basho/riak/client/api/commands/mapreduce/BinaryValueSerializer.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonSerializer;
+import com.fasterxml.jackson.databind.SerializerProvider;
+
+import java.io.IOException;
+
+/**
+ * Serializes a {@link BinaryValue} to a UTF-8 string for the Riak MapReduce JSON format.
+ */
+public class BinaryValueSerializer extends JsonSerializer<BinaryValue>
```
Comment (srgg) [inline src/main/java/com/basho/riak/client/api/commands/mapreduce/BinaryValueSerializer.java:13]: Should it be registered somehow to be used by default without a necessity to be registered explicitly by user?

**UPD:**
Found that it is registered in MapReduce.writeSpec, but what about pulling all JsonGenerator creation related stuff into separate static function that will be reused in the Tests?

**Comment by christophermancini** [comment]
👍 Tests pass, the code looks reasonable but I am having a hard time following the changes. :(

File: src/main/java/com/basho/riak/client/api/commands/mapreduce/BinaryValueSerializer.java (hunk 1 above)
Comment (alexmoore) [inline src/main/java/com/basho/riak/client/api/commands/mapreduce/BinaryValueSerializer.java:13]: Done with b101124.

**Comment by srgg** [comment]
👍

**Pull request merged**